Add PSContentPath Infrastructure

jshigetomi opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This pull request introduces new user-facing cmdlets for managing the PowerShell content path, refactors how personal module and configuration paths are determined, and improves cross-platform consistency for configuration storage. The most significant changes are the addition of `Get-PSContentPath` and `Set-PSContentPath` cmdlets, updates to the logic for resolving user content/config paths (especially on Windows), and the adoption of the content path as the basis for the personal module path.

**Platform-specific content/config path handling:**
- Introduced `DefaultPSContentDirectory` and `LocalAppDataPSContentDirectory` properties in `CorePsPlatform.cs` to standardize the location of PowerShell content data across platforms. On Windows, `Documents\PowerShell` is still preferred for PSContentPath , with a fallback for legacy locations. This was to avoid a breaking change and just make this PR an introduction for the infrastructure.
- Refactored `PowerShellConfig` to select the user config directory/file based on the presence of existing config files, ensuring backward compatibility while defaulting to the new location for new installs.

**Personal module path changes:**
- Changed the logic for determining the personal module path to use the PS content path (from `GetPSContentPath()`), rather than hardcoded locations like `My Documents` or XDG directories.
- Updated the module path resolution to fall back to the new personal module path if not explicitly set in config.

**Constants and documentation improvements:**
- Added a constant for the PS user content path config key in `ModuleIntrinsics.cs`.
- Clarified documentation for module constraint matching.

**New cmdlets for PSContentPath management:**
- Added `Get-PSContentPath` and `Set-PSContentPath` cmdlets, allowing users to query and configure the PowerShell content path with validation logic.
- Registered these cmdlets in the core session state, making them available by default. [[1]](diffhunk://#diff-ad3fd038bf04bb095940b4786d074a68d6c310f623bdd7cfe8f60fc7775a495aR5479) [[2]](diffhunk://#diff-ad3fd038bf04bb095940b4786d074a68d6c310f623bdd7cfe8f60fc7775a495aR5502)

### `Get-PSContentPath`

Gets the current PowerShell user content path (where modules, scripts, help, and profiles are stored). Returns a `System.IO.DirectoryInfo` object with a `ConfigFile` NoteProperty attached, which the default list formatter displays.

```powershell
# Get the current content path
(Get-PSContentPath).FullName

# Use in string interpolation
"Modules are stored in $(Get-PSContentPath)\Modules"

# Get the config file location
(Get-PSContentPath).ConfigFile
```

### `Set-PSContentPath`

Sets a custom content path, persisted in `powershell.config.json`. The confirm impact is `High`, so users are prompted for confirmation by default.

| Parameter | Description |
|-----------|-------------|
| `-Path <string>` | Absolute path for the content directory (env vars expanded) |
| `-Default` | Resets to platform default |
| `-WhatIf` / `-Confirm` | ShouldProcess support (prompts by default due to High impact) |

```powershell
PS> Set-PSContentPath -Path "D:\PowerShell"

Confirm
Are you sure you want to perform this action?
...
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"):

PS> Set-PSContentPath -Path "$env:OneDrive\PowerShell"
PS> Set-PSContentPath -Default
```

### `$PSUserContentPath` Variable

A readonly automatic variable exposing the current content path. Users cannot modify it directly—they get a helpful error directing them to use `Set-PSContentPath`.

```powershell
PS> $PSUserContentPath
C:\Users\user\Documents\PowerShell

PS> $PSUserContentPath = "D:\Other"
InvalidOperation: Cannot modify the readonly variable 'PSUserContentPath'. 
Use Set-PSContentPath to configure this value.
```

---

## PR Context

This change represents a major milestone for the PowerShell ecosystem—one that has been long requested by the community, and discussed extensively. The feature originated from community feedback in https://github.com/PowerShell/PowerShell/issues/15552, aligns with related work in https://github.com/PowerShell/PSResourceGet/pull/1912, and was shaped through detailed specifications in https://github.com/PowerShell/PowerShell-RFC/pull/388.

This feature lays the foundation for flexible, user-configurable PowerShell content storage, addressing scenarios such as roaming profiles and containerized environments.

By enabling customization of content paths, we empower users to adapt PowerShell to modern deployment and security requirements.

Partially Addresses: #26121 #27370 #24274

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge. If this PR is a work in progress, please open this as a [Draft Pull Request and mark it as Ready to Review when it is ready to merge](https://docs.github.com/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests).
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): PSContentPath
- **User-facing changes**
  - [ ] Not Applicable
  - **OR**
  - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/12827
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

Commit:
Testing

Commit:
Update PSModulePath

jshigetomi committed:
Add Experimental feature for PSContent

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

jshigetomi committed:
Removed unused variables

jshigetomi committed:
Switch to default PSContentPath LOCALAPPData, cmdlets added

jshigetomi committed:
Add lazy migration

jshigetomi committed:
Add null checks incase PSUserContentPath fails to get any value

jshigetomi committed:
Able to use expanded environmental variables

jshigetomi committed:
Remove help URI

jshigetomi committed:
Reassign perUserConfigDirectory if experimental feature is enabled whether or not migration took place

jshigetomi committed:
Move migration to GetPSContentPath API, added safety fallbacks to default PSContentPath location

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

iSazonov commented:
If user intention is to migrate content to new location I believe the new cmdlets should help them.
I'd like to get not only current content path but size too `Get-PSContentPath -MeasureSize`.

jshigetomi commented:
@kilasuit - Thanks for calling me out on that. Definitely intended this as a draft.

jshigetomi committed:
Add tests

jshigetomi committed:
Merge branch 'PSModulePathFix' of https://github.com/jshigetomi/PowerShell into PSModulePathFix

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

jshigetomi committed:
Add initial test cases for PSContentPath

jshigetomi commented on src/System.Management.Automation/engine/Configuration/GetSetPSContentPathCommand.cs:
```diff
@@ -0,0 +1,169 @@
+        /// <summary>
+        /// ProcessRecord method of this cmdlet.
+        /// </summary>
+        protected override void ProcessRecord()
```
Good catch. Done

jshigetomi committed:
Fix expanding env variable test

jshigetomi committed:
Separate commands for env var test

jshigetomi commented:
@ISazonov Get-PSContentPath is only available when the PSContentPath experimental feature is enabled. When enabled, the content path defaults to LocalAppData rather than the OneDrive-synced Documents folder. This means -MeasureSize would only be useful if a user explicitly sets their content path back to the OneDrive location, which defeats the purpose of the feature.

Should we consider making Get-PSContentPath available outside the experimental feature gate? That way users could measure their current Documents\PowerShell directory size even without enabling the feature.

jshigetomi commented:
@iSazonov 
> 
> Gradually, all the resources will end up in new folders.
> 
I am concerned about this step. I believe users will not know about this feature and not enable the experimental feature to begin the gradual moving of resources. 

To help with this transition we can eventually disable the old path by default like you said. But this would out of the box break users who haven't migrated, which I suspect could be the majority. The fix in that situation would be to Set-PSContentPath -Path "ONEDRIVE".

Maybe we can start off the PSContentPath feature to be OneDrive out of the box. This way it doesn't break users but allows them to customize their path to a desired location. They can then use something like the following to set it easily in default known locations. 

```pwsh
Set-PSContentPath -LocalAppData  ## These parameters don't exist yet but open to adding them
Set-PSContentPath -OneDrive
```

jshigetomi commented:
@kilasuit 
> I think the plan would be that this is all contained within 7.6.x as not to drag this out for more years but that's something that @jshigetomi & team can confirm on, but that was my expectation at least
I think the plan as of now is to introduce it in 7.7 preview and stabilize it in 7.7 rc. Definitely do not want to drag this on for more years than it already has.

iSazonov commented:
> I am concerned about this step. I believe users will not know about this feature and not enable the experimental feature to begin the gradual moving of resources.

Notice, experimental feature is enabled by default. But really we could avoid experimental feature at all and make it as regular.

I believe that any migration involving copying folders, enabling or disabling any settings, will always cause massive problems for users.

Currently we look one standard folder, then second and so on. If we put the new APPDATA folder before Documents folder the new folder will have higher priority and will work transparently. So we have no need to make it as experimental. For better UX install cmdlet could warn about this new default and maybe explicitly allow old behavior (what if anybody uses OneDrive cloud replication?).

jshigetomi commented:
@iSazonov 
In this PR I wrote it so that it disables PSContentPath by default.

Interesting idea on making it non-experimental and putting AppData before OneDrive. I will bring it up in our design meeting.

I do see a need to remove OneDrive from the path entirely and users need a way to do that.

jshigetomi commented:
@iSazonov 
> I do see a need to remove OneDrive from the path entirely and users need a way to do that.

By this I meant to give an option to remove OneDrive. One of the main goals of this is to provide a way for users to avoid syncing with OneDrive. 

I do like the soft path you said. In that scenario Set-PSContentPath -Remove could remove OneDrive as a location for pwsh.

kilasuit commented:
Whilst I can also help, in the area's I've mentioned, I know there are many others that could do too and it would be good for others to step up and take the lead on their own experiments as to enable us to get this in sooner rather than later. Of course, I realise that this has to be correctly prioritised by all that wanna get involved in making this happen, with their other workloads. But I really would like it, if, we could make it happen before April next year.

jshigetomi commented:
After discussing with the PowerShell team I think the best approach is to make this not an experimental feature.
It will only introduce the new infrastructure to customize PowerShell content path and point it at OneDrive by default to prevent breaking users. For most users nothing would change and for those users that want to relocate, they can do so.

The benefits of this approach include the following:
- There is no behavioral change on the users' part so this is an infrastructure change not an experimental feature. 
- Removing the experimental feature aspect of this will simplify the code significantly. 
- Keeps the opt-in aspect of this feature where users decide whether to switch locations.

The plan is now to introduce an experimental feature in the next PR that sets the default to LocalAppData.

jshigetomi committed:
Initial redo to remove experimental feature aspect and point to OneDrive by default

jshigetomi committed:
Switch default PSContentPath variable to OneDrive

jshigetomi committed:
Remove old PSUserContentPath code

jshigetomi committed:
Fix help test and add Get/Set to cmdlet test

jshigetomi committed:
Update tab completion tests with PSContentPath resolution

jshigetomi committed:
Add warning message in tabcompletion tests

jshigetomi committed:
Remove changes to build script

jshigetomi committed:
Add Move-PSContentPath cmdlet

jshigetomi committed:
Skip Move-PSContentPath in HelpSystem tests for now

jshigetomi committed:
Separate out fallback resolution for UNIX/Windows

jshigetomi committed:
Add Move-PSContent to cmdlets list and clean up code

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

jshigetomi committed:
Add back switch-process

jshigetomi committed:
Merge branch 'PSModulePathFix' of https://github.com/jshigetomi/PowerShell into PSModulePathFix

jshigetomi commented:
@iSazonov @kilasuit 
Added a cmdlet Move-PSContent that will move or copy things over to the new location.
Move-PSContentPath -Destination is limited to only 2 destinations, OneDrive or LocalAppData. This is to ensure PowerShell can find the user config path on startup. 
Added fallback logic for the user config to look first in LocalAppData then in OneDrive.
I also added the -Size parameter for Get-PSContentPath to see the size of the move.

jshigetomi commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -144,6 +164,66 @@ internal string GetModulePath(ConfigScope scope)
+        /// If not configured, returns the default OneDrive location (Documents\PowerShell) without creating the config file.
+        /// This ensures PowerShell works on read-only file systems and avoids creating unnecessary files.
+        /// </summary>
+        /// <returns>The configured PSContentPath if found, otherwise the default OneDrive location (never null).</returns>
```
done

jshigetomi commented on src/System.Management.Automation/engine/Configuration/PSContentPathCommands.cs:
```diff
@@ -0,0 +1,453 @@
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 0)]
+        [ValidateSet("OneDrive", "LocalAppData")]
+        public string Path { get; set; }
```
Move-Item and Copy-Item use -Path -Destination

jshigetomi commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -16,17 +16,18 @@ internal static class HelpUtils
         /// <summary>
         /// Get the path to $HOME.
```
done

jshigetomi commented on src/System.Management.Automation/help/HelpUtils.cs:
```diff
@@ -16,17 +16,18 @@ internal static class HelpUtils
-#endif
-                userHomeHelpPath = Path.Combine(userScopeRootPath, "Help");
+                userHomeHelpPath = expectedPath;
             }
```
Not taking this

jshigetomi commented on src/System.Management.Automation/engine/Configuration/PSContentPathCommands.cs:
```diff
@@ -0,0 +1,453 @@
+            }
+
+            string action = Copy ? "Copy" : "Move";
+            string sourcePattern = System.IO.Path.Combine(sourcePath, "*");
```
done

Commit:
Changes from Copilot review

Commit:
Fix Set-PSContentPath logic, add summary for Get-PSContentPath

jshigetomi commented:
@iSazonov Just an FYI this feature is not planned to be supported in PowerShellGet unfortunately. Only for the latest PSResourceGet. 

And that's a good point. This shows that something like this has been done before and we can follow that example.

jshigetomi commented on src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs:
```diff
@@ -19,6 +19,7 @@ namespace System.Management.Automation
     internal static class Constants
     {
         public const string PSModulePathEnvVar = "PSModulePath";
+        public const string PSUserContentPathEnvVar = "PSUserContentPath";
```
Make this read only pwsh varaible, Error message "If you want to set this use Set-PSContentPath"
Remove current default PSModulePath, they can add it on their profile if they want it.

jshigetomi committed:
Add PSUserContentPath as a pwsh variable, Remove Move-PSContentPath

jshigetomi committed:
Make LocalAppData default config file location

jshigetomi committed:
Add PSUSerContentVariable class to emit custom error message, Fix tests after moving config file to LocalAppData

jshigetomi commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -85,14 +85,34 @@ private PowerShellConfig()
+            string localAppDataConfig = Platform.LocalAppDataPSContentDirectory;
+            string localAppDataConfigFile = Path.Combine(localAppDataConfig, ConfigFileName);
+
+            if (File.Exists(localAppDataConfigFile))
```
done

jshigetomi committed:
Point test fixutre at new locaiton for config LocalAppData

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

jshigetomi committed:
Add -ConfigFile parameter to Get-PSContentPath, -WhatIf

jshigetomi committed:
Add both paths to PSModulePath

jshigetomi commented:
@iSazonov 
>  So I'm sure the only thing that needs to be done here is to update the engine to process the new path in the first place and that's it.

I agree. That's what we are doing for Windows PowerShell in pwsh. We kept both and placed pwsh first.

jshigetomi committed:
Rename to PSContentCommands

SteveL-MSFT commented:
The @PowerShell/wg-powershell-cmdlets discussed the cmdlets.  For `Get-PSContentPath` we don't need the `-Size` switch nor the `-ConfigFile` switch.  Size information shouldn't be returned.  For `-ConfigFile`, the cmdlet should return a `System.IO.DirectoryInfo` object, but add a `NoteProperty` that contains the `ConfigFile` and the default formatter for list view can display that.  For `Set-PSContentPath`, `-Reset` should be called `-Default` and the confirm impact for this cmdlet should be `High` so that it prompts by default.

jshigetomi committed:
Add config file as a note property and rename -reset parameter to default

jshigetomi committed:
Change impact for Set-PSContentPath to High

Commit:
Copilot review fixes

Commit:
Add in necessary directives

Commit:
- Add null check for GetPersonalModulePath.
- Added tests for automatic variable.
- Added tests for -default parameter
- added tests for legacty documents path
-Added test for deduplication of PSModulePath

Commit:
Remove readonly for PSUserContentPath and rely on setter to throw a helpful error
Fix tests

Commit:
Remove tests for helpful error message

Commit:
Remove deduplicatation test because tests inject module path

jshigetomi committed:
Merge branch 'master' into PSModulePathFix

dn-jmp commented:
Hello! I am looking forward to this implementation and I am just curious if it has been slated for any particular release milestones. Do you think it'll make it into 7.7.0?

StartAutomating commented:
Please also provide Push/Pop-PSContentPath.

_especially_ because you're planning on using a fallback stack anyways.

This would be _incredibly_ useful when working in certain contexts, like git worktreees.

jshigetomi commented:
@dn-jmp The goal is to make it into 7.7

StartAutomating commented:
One other curveball to potentially support would be having no content path.

Why?  Lockdown scenarios and isolation testing.

If I wanted to isolate a particular load error, I would want to ensure that no content is loaded from other directories.

Similarly, if I wanted to lock down a particular session so that it cannot load other things, I would want to load whatever is required, change the content paths, and change the language mode. If there was no acceptable location to load content from, this would help harden the surface area in these scenarios.

jshigetomi commented:
@StartAutomating the no content path scenario is interesting. Right now, the content is made available via the PSModulePath variable, and this PR just provides infrastructure to add the defined content path to PSModulePath.

The scope of this work would be beyond just the user level.

StartAutomating commented:
> James Brundage (@StartAutomating) the no content path scenario is interesting. Right now, the content is made available via the PSModulePath variable, and this PR just provides infrastructure to add the defined content path to PSModulePath.
> 
> The scope of this work would be beyond just the user level.

Fine to know.

If we're just manipulating PSModulePath, then _definitely_ Push/Pop-PSContentPath.

Also, if we're manipulating PSModulePath, for discoverability I'd highly suggest `Get/Set/Push/Pop-PSModulePath`
